File: TrabalhoPARTE3/Parser.cpp


#include <wchar.h>
#include "Parser.h"
#include "Scanner.h"




void Parser::SynErr(int n) {
	if (errDist >= minErrDist) errors->SynErr(la->line, la->col, n);
	errDist = 0;
}

void Parser::SemErr(const wchar_t* msg) {
	if (errDist >= minErrDist) errors->Error(t->line, t->col, msg);
	errDist = 0;
}

void Parser::Get() {
	for (;;) {
		t = la;
		la = scanner->Scan();
		if (la->kind <= maxT) { ++errDist; break; }

		if (dummyToken != t) {
			dummyToken->kind = t->kind;
			dummyToken->pos = t->pos;
			dummyToken->col = t->col;
			dummyToken->line = t->line;
			dummyToken->next = NULL;
			coco_string_delete(dummyToken->val);
			dummyToken->val = coco_string_create(t->val);
			t = dummyToken;
		}
		la = t;
	}
}

void Parser::Expect(int n) {
	if (la->kind==n) Get(); else { SynErr(n); }
}

void Parser::ExpectWeak(int n, int follow) {
	if (la->kind == n) Get();
	else {
		SynErr(n);
		while (!StartOf(follow)) Get();
	}
}

bool Parser::WeakSeparator(int n, int syFol, int repFol) {
	if (la->kind == n) {Get(); return true;}
	else if (StartOf(repFol)) {return false;}
	else {
		SynErr(n);
		while (!(StartOf(syFol) || StartOf(repFol) || StartOf(0))) {
			Get();
		}
		return StartOf(syFol);
	}
}

void Parser::CLC() {
		Expect(3 /* "begin" */);
		L();
		Expect(4 /* "end." */);
}

void Parser::L() {
		if (StartOf(1)) {
			C();
			Expect(5 /* ";" */);
			L();
		} else if (la->kind == 6 /* "showSymbolTable" */) {
			Get();
			Expect(7 /* "(" */);
			Expect(8 /* ")" */);
			Expect(5 /* ";" */);
		} else if (la->kind == 9 /* "showSymbolTree" */) {
			Get();
			Expect(7 /* "(" */);
			Expect(8 /* ")" */);
			Expect(5 /* ";" */);
		} else SynErr(28);
}

void Parser::C() {
		if (la->kind == _id) {
			Get();
			Expect(10 /* ":=" */);
			E();
		} else if (la->kind == 11 /* "input" */) {
			Get();
			Expect(7 /* "(" */);
			Expect(_id);
			Expect(8 /* ")" */);
		} else if (la->kind == 12 /* "print" */) {
			Get();
			Expect(7 /* "(" */);
			Expect(_id);
			Expect(8 /* ")" */);
		} else if (la->kind == 23 /* "do" */) {
			DOWHILE();
		} else SynErr(29);
}

void Parser::E() {
		T();
		while (la->kind == 13 /* "+" */ || la->kind == 14 /* "-" */) {
			if (la->kind == 13 /* "+" */) {
				Get();
			} else {
				Get();
			}
			T();
		}
}

void Parser::DOWHILE() {
		Expect(23 /* "do" */);
		Expect(24 /* "{" */);
		L();
		Expect(25 /* "}" */);
		Expect(26 /* "while" */);
		Expect(7 /* "(" */);
		COMPARA();
		Expect(8 /* ")" */);
		Expect(5 /* ";" */);
}

void Parser::T() {
		F();
		while (la->kind == 15 /* "*" */ || la->kind == 16 /* "/" */) {
			if (la->kind == 15 /* "*" */) {
				Get();
			} else {
				Get();
			}
			F();
		}
}

void Parser::F() {
		if (la->kind == _id) {
			Get();
		} else if (la->kind == _num) {
			Get();
		} else if (la->kind == 7 /* "(" */) {
			Get();
			E();
			Expect(8 /* ")" */);
		} else SynErr(30);
}

void Parser::COMPARA() {
		if (StartOf(1)) {
			C();
			Expect(17 /* "==" */);
			C();
		} else if (StartOf(1)) {
			C();
			Expect(18 /* ">=" */);
			C();
		} else if (StartOf(1)) {
			C();
			Expect(19 /* "<=" */);
			C();
		} else if (StartOf(1)) {
			C();
			Expect(20 /* ">" */);
			C();
		} else if (StartOf(1)) {
			C();
			Expect(21 /* "<" */);
			C();
		} else if (StartOf(1)) {
			C();
			Expect(22 /* "!=" */);
			C();
		} else SynErr(31);
}




// If the user declared a method Init and a mehtod Destroy they should
// be called in the contructur and the destructor respctively.
//
// The following templates are used to recognize if the user declared
// the methods Init and Destroy.

template<typename T>
struct ParserInitExistsRecognizer {
	template<typename U, void (U::*)() = &U::Init>
	struct ExistsIfInitIsDefinedMarker{};

	struct InitIsMissingType {
		char dummy1;
	};
	
	struct InitExistsType {
		char dummy1; char dummy2;
	};

	// exists always
	template<typename U>
	static InitIsMissingType is_here(...);

	// exist only if ExistsIfInitIsDefinedMarker is defined
	template<typename U>
	static InitExistsType is_here(ExistsIfInitIsDefinedMarker<U>*);

	enum { InitExists = (sizeof(is_here<T>(NULL)) == sizeof(InitExistsType)) };
};

template<typename T>
struct ParserDestroyExistsRecognizer {
	template<typename U, void (U::*)() = &U::Destroy>
	struct ExistsIfDestroyIsDefinedMarker{};

	struct DestroyIsMissingType {
		char dummy1;
	};
	
	struct DestroyExistsType {
		char dummy1; char dummy2;
	};

	// exists always
	template<typename U>
	static DestroyIsMissingType is_here(...);

	// exist only if ExistsIfDestroyIsDefinedMarker is defined
	template<typename U>
	static DestroyExistsType is_here(ExistsIfDestroyIsDefinedMarker<U>*);

	enum { DestroyExists = (sizeof(is_here<T>(NULL)) == sizeof(DestroyExistsType)) };
};

// The folloing templates are used to call the Init and Destroy methods if they exist.

// Generic case of the ParserInitCaller, gets used if the Init method is missing
template<typename T, bool = ParserInitExistsRecognizer<T>::InitExists>
struct ParserInitCaller {
	static void CallInit(T *t) {
		// nothing to do
	}
};

// True case of the ParserInitCaller, gets used if the Init method exists
template<typename T>
struct ParserInitCaller<T, true> {
	static void CallInit(T *t) {
		t->Init();
	}
};

// Generic case of the ParserDestroyCaller, gets used if the Destroy method is missing
template<typename T, bool = ParserDestroyExistsRecognizer<T>::DestroyExists>
struct ParserDestroyCaller {
	static void CallDestroy(T *t) {
		// nothing to do
	}
};

// True case of the ParserDestroyCaller, gets used if the Destroy method exists
template<typename T>
struct ParserDestroyCaller<T, true> {
	static void CallDestroy(T *t) {
		t->Destroy();
	}
};

void Parser::Parse() {
	t = NULL;
	la = dummyToken = new Token();
	la->val = coco_string_create(L"Dummy Token");
	Get();
	CLC();
	Expect(0);
}

Parser::Parser(Scanner *scanner) {
	maxT = 27;

	ParserInitCaller<Parser>::CallInit(this);
	dummyToken = NULL;
	t = la = NULL;
	minErrDist = 2;
	errDist = minErrDist;
	this->scanner = scanner;
	errors = new Errors();
}

bool Parser::StartOf(int s) {
	const bool T = true;
	const bool x = false;

	static bool set[2][29] = {
		{T,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x},
		{x,T,x,x, x,x,x,x, x,x,x,T, T,x,x,x, x,x,x,x, x,x,x,T, x,x,x,x, x}
	};



	return set[s][la->kind];
}

Parser::~Parser() {
	ParserDestroyCaller<Parser>::CallDestroy(this);
	delete errors;
	delete dummyToken;
}

Errors::Errors() {
	count = 0;
}

void Errors::SynErr(int line, int col, int n) {
	wchar_t* s;
	switch (n) {
			case 0: s = coco_string_create(L"EOF expected"); break;
			case 1: s = coco_string_create(L"id expected"); break;
			case 2: s = coco_string_create(L"num expected"); break;
			case 3: s = coco_string_create(L"\"begin\" expected"); break;
			case 4: s = coco_string_create(L"\"end.\" expected"); break;
			case 5: s = coco_string_create(L"\";\" expected"); break;
			case 6: s = coco_string_create(L"\"showSymbolTable\" expected"); break;
			case 7: s = coco_string_create(L"\"(\" expected"); break;
			case 8: s = coco_string_create(L"\")\" expected"); break;
			case 9: s = coco_string_create(L"\"showSymbolTree\" expected"); break;
			case 10: s = coco_string_create(L"\":=\" expected"); break;
			case 11: s = coco_string_create(L"\"input\" expected"); break;
			case 12: s = coco_string_create(L"\"print\" expected"); break;
			case 13: s = coco_string_create(L"\"+\" expected"); break;
			case 14: s = coco_string_create(L"\"-\" expected"); break;
			case 15: s = coco_string_create(L"\"*\" expected"); break;
			case 16: s = coco_string_create(L"\"/\" expected"); break;
			case 17: s = coco_string_create(L"\"==\" expected"); break;
			case 18: s = coco_string_create(L"\">=\" expected"); break;
			case 19: s = coco_string_create(L"\"<=\" expected"); break;
			case 20: s = coco_string_create(L"\">\" expected"); break;
			case 21: s = coco_string_create(L"\"<\" expected"); break;
			case 22: s = coco_string_create(L"\"!=\" expected"); break;
			case 23: s = coco_string_create(L"\"do\" expected"); break;
			case 24: s = coco_string_create(L"\"{\" expected"); break;
			case 25: s = coco_string_create(L"\"}\" expected"); break;
			case 26: s = coco_string_create(L"\"while\" expected"); break;
			case 27: s = coco_string_create(L"??? expected"); break;
			case 28: s = coco_string_create(L"invalid L"); break;
			case 29: s = coco_string_create(L"invalid C"); break;
			case 30: s = coco_string_create(L"invalid F"); break;
			case 31: s = coco_string_create(L"invalid COMPARA"); break;

		default:
		{
			wchar_t format[20];
			coco_swprintf(format, 20, L"error %d", n);
			s = coco_string_create(format);
		}
		break;
	}
	wprintf(L"-- line %d col %d: %ls\n", line, col, s);
	coco_string_delete(s);
	count++;
}

void Errors::Error(int line, int col, const wchar_t *s) {
	wprintf(L"-- line %d col %d: %ls\n", line, col, s);
	count++;
}

void Errors::Warning(int line, int col, const wchar_t *s) {
	wprintf(L"-- line %d col %d: %ls\n", line, col, s);
}

void Errors::Warning(const wchar_t *s) {
	wprintf(L"%ls\n", s);
}

void Errors::Exception(const wchar_t* s) {
	wprintf(L"%ls", s); 
	exit(1);
}

File: TrabalhoPARTE2/Parser.java


public class Parser {
	public static final int _EOF = 0;
	public static final int _id = 1;
	public static final int _num = 2;
	public static final int maxT = 17;

	static final boolean _T = true;
	static final boolean _x = false;
	static final int minErrDist = 2;

	public Token t;    // last recognized token
	public Token la;   // lookahead token
	int errDist = minErrDist;
	
	public Scanner scanner;
	public Errors errors;

	public void showProduction(String s) {
  	System.out.println("Token lido: "+s);
}

String s = "";



	public Parser(Scanner scanner) {
		this.scanner = scanner;
		errors = new Errors();
	}

	void SynErr (int n) {
		if (errDist >= minErrDist) errors.SynErr(la.line, la.col, n);
		errDist = 0;
	}

	public void SemErr (String msg) {
		if (errDist >= minErrDist) errors.SemErr(t.line, t.col, msg);
		errDist = 0;
	}
	
	void Get () {
		for (;;) {
			t = la;
			la = scanner.Scan();
			if (la.kind <= maxT) {
				++errDist;
				break;
			}

			la = t;
		}
	}
	
	void Expect (int n) {
		if (la.kind==n) Get(); else { SynErr(n); }
	}
	
	boolean StartOf (int s) {
		return set[s][la.kind];
	}
	
	void ExpectWeak (int n, int follow) {
		if (la.kind == n) Get();
		else {
			SynErr(n);
			while (!StartOf(follow)) Get();
		}
	}
	
	boolean WeakSeparator (int n, int syFol, int repFol) {
		int kind = la.kind;
		if (kind == n) { Get(); return true; }
		else if (StartOf(repFol)) return false;
		else {
			SynErr(n);
			while (!(set[syFol][kind] || set[repFol][kind] || set[0][kind])) {
				Get();
				kind = la.kind;
			}
			return StartOf(syFol);
		}
	}
	
	void CLC() {
		Expect(3);
		showProduction(t.val);
		L();
		Expect(4);
		showProduction(t.val);
	}

	void L() {
		if (la.kind == 1 || la.kind == 11 || la.kind == 12) {
			C();
			Expect(5);
			L();
		} else if (la.kind == 6) {
			Get();
			Expect(7);
			Expect(8);
			Expect(5);
		} else if (la.kind == 9) {
			Get();
			showProduction(t.val);
			Expect(7);
			Expect(8);
			Expect(5);
		} else SynErr(18);
	}

	void C() {
		if (la.kind == 1) {
			Get();
			Expect(10);
			E();
			showProduction(t.val);
		} else if (la.kind == 11) {
			Get();
			showProduction(t.val);
			Expect(7);
			Expect(1);
			Expect(8);
		} else if (la.kind == 12) {
			Get();
			showProduction(t.val);
			Expect(7);
			Expect(1);
			Expect(8);
		} else SynErr(19);
	}

	void E() {
		T();
		while (la.kind == 13 || la.kind == 14) {
			if (la.kind == 13) {
				Get();
			} else {
				Get();
			}
			T();
		}
		showProduction(t.val);
	}

	void T() {
		F();
		while (la.kind == 15 || la.kind == 16) {
			if (la.kind == 15) {
				Get();
			} else {
				Get();
			}
			F();
		}
		showProduction(t.val);
	}

	void F() {
		if (la.kind == 1) {
			Get();
			showProduction(t.val);
		} else if (la.kind == 2) {
			Get();
			showProduction(t.val);
		} else if (la.kind == 7) {
			Get();
			E();
			Expect(8);
		} else SynErr(20);
	}



	public void Parse() {
		la = new Token();
		la.val = "";		
		Get();
		CLC();
		Expect(0);

	}

	private static final boolean[][] set = {
		{_T,_x,_x,_x, _x,_x,_x,_x, _x,_x,_x,_x, _x,_x,_x,_x, _x,_x,_x}

	};
} // end Parser


class Errors {
	public int count = 0;                                    // number of errors detected
	public java.io.PrintStream errorStream = System.out;     // error messages go to this stream
	public String errMsgFormat = "-- line {0} col {1}: {2}"; // 0=line, 1=column, 2=text
	
	protected void printMsg(int line, int column, String msg) {
		StringBuffer b = new StringBuffer(errMsgFormat);
		int pos = b.indexOf("{0}");
		if (pos >= 0) { b.delete(pos, pos+3); b.insert(pos, line); }
		pos = b.indexOf("{1}");
		if (pos >= 0) { b.delete(pos, pos+3); b.insert(pos, column); }
		pos = b.indexOf("{2}");
		if (pos >= 0) b.replace(pos, pos+3, msg);
		errorStream.println(b.toString());
	}
	
	public void SynErr (int line, int col, int n) {
		String s;
		switch (n) {
			case 0: s = "EOF expected"; break;
			case 1: s = "id expected"; break;
			case 2: s = "num expected"; break;
			case 3: s = "\"begin\" expected"; break;
			case 4: s = "\"end.\" expected"; break;
			case 5: s = "\";\" expected"; break;
			case 6: s = "\"showSymbolTable\" expected"; break;
			case 7: s = "\"(\" expected"; break;
			case 8: s = "\")\" expected"; break;
			case 9: s = "\"showSymbolTree\" expected"; break;
			case 10: s = "\":=\" expected"; break;
			case 11: s = "\"input\" expected"; break;
			case 12: s = "\"print\" expected"; break;
			case 13: s = "\"+\" expected"; break;
			case 14: s = "\"-\" expected"; break;
			case 15: s = "\"*\" expected"; break;
			case 16: s = "\"/\" expected"; break;
			case 17: s = "??? expected"; break;
			case 18: s = "invalid L"; break;
			case 19: s = "invalid C"; break;
			case 20: s = "invalid F"; break;
			default: s = "error " + n; break;
		}
		printMsg(line, col, s);
		count++;
	}

	public void SemErr (int line, int col, String s) {	
		printMsg(line, col, s);
		count++;
	}
	
	public void SemErr (String s) {
		errorStream.println(s);
		count++;
	}
	
	public void Warning (int line, int col, String s) {	
		printMsg(line, col, s);
	}
	
	public void Warning (String s) {
		errorStream.println(s);
	}
} // Errors


class FatalError extends RuntimeException {
	public static final long serialVersionUID = 1L;
	public FatalError(String s) { super(s); }
}
